lab4/src/main/java/agh/ics/oop/SimulationEngine.java

```diff
@@ -0,0 +1,37 @@
+package agh.ics.oop;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class SimulationEngine implements IEngine{
+    private final MoveDirection[] moveDirections;
+    private final IWorldMap map;
+    private final Vector2d[] animalsVectors;
+    private final List<Animal> animalsOnMap = new ArrayList<>();
+
+    public SimulationEngine(MoveDirection[] moveDirections, IWorldMap map, Vector2d[] animalsVectors){
+        this.moveDirections = moveDirections;
+        this.map = map;
+        this.animalsVectors = animalsVectors;
+    }
+    private int addAndReturnNumberOfAnimals(){
+
+        int numberOfAnimalsOnMap = 0;
+
+        for (Vector2d animalVector : animalsVectors){
+            Animal animalToInsert = new Animal(this.map, animalVector);
+            if(this.map.place(animalToInsert)){
+                this.animalsOnMap.add(animalToInsert);
+                numberOfAnimalsOnMap++;
+            }
+        }
+        return numberOfAnimalsOnMap;
+    }
+    public void run(){
+
+        int numberOfAnimalsOnMap = this.addAndReturnNumberOfAnimals();
```

Comment:
metoda `addAndReturnNumberOfAnimals()` powinna być raczej wołana raz, w konstrukturze, bo w tym momencie jak ktoś zrobić kilka razy `run()` to będzie za każdym razem dokładać kolejne zwierzątka do mapy (na początku będzie ich N, potem 2N, potem 3N itd).